src/debug.s

```diff
@@ -1,5 +1,317 @@
+.org 0807E4B6h
+.area 0807E4C8h-., 0
+    ldr     r4, =@DebugMenuDrawHealthAndAmmoHighjack
+    mov     pc, r4
+    .pool
+.endarea
+.definelabel @ReturnFromDebugMenuDrawingHighjack, org()+1
+
+
+; Allow speeding or slowing noclip speed
+.org 0800E6F0h
+.area 0800E702h-., 0
+    ldr     r0, =@NoClipSpeedHighjack
+    mov     pc, r0
+    .pool
+.endarea
+.definelabel @ReturnFromNoClipSpeedHighjack, org()+1
+
+
+.autoregion
+    .align 2
+.func EnableDebug
+    push    { r0-r1, lr }
+    ldr     r1, =DebugFlag
+    mov     r0, #01h
+    strb    r0, [r1]
+    pop     { r0-r1, pc }
+    .pool
+.endfunc
+
+/* NOTES
+    Hold Button_L and Button_R, Press Button_Select to enable No Clip.
+    While in No Clip mode:
+    * hold Button_R to move quickly
+    * hold Button_L to move slowly
+    * press Button_B to give yourself a speedbooster timer
+    * press Button_Start to center the camera on Samus and disable room scrolls.
+      This lets you see all GFX of the room, including the normally hidden borders.
+    * press Button_Select to exit No Clip mode.
+    * Samus cannot be damaged.
+    * Samus cannot interact with level geometry, door transitions, or sprites.
+
+*/
+    .align 2
+@CheckEnableNoClip:
+    ldr     r0, =DebugFlag
+    ldrb    r0, [r0]
+    cmp     r0, #00h
+    beq     @@if_false
+    ldr     r0, =ToggleInput
+    ldrh    r0, [r0]
+    mov     r1, #1 << Button_Select
+    and     r0, r1                  ; If Button_Select Pressed ...
+    cmp     r0, #00h
+    beq     @@if_false
+    ldr     r0, =HeldInput
+    ldrh    r0, [r0]
+    mov     r1, #(1 << (Button_L | Button_R)) >> 4
+    lsl     r1, #04h
+    mov     r2, r1
+    and     r0, r2                  ; ... while holding Button_L and Button_R ...
+    cmp     r0, r1
+    bne     @@if_false
+
+@@if_true:
+    ldr     r0, =TimeStopTimer
+    ldrh    r0, [r0]
+    cmp     r0, #00h
+    bne     @@if_false
+    ldr     r0, =SubGameMode1
+    mov     r1, #06h                 ; ... then enable no-clip mode.
+    strh    r1, [r0]
+    b       @@return
+
+@@if_false: ; Restore vanilla functionality
+    ldr     r0, =ToggleInput
+    ldr     r1, =ButtonAssignments
+    ldrh    r2, [r0]
+    ldrh    r0, [r1, #ButtonAssignments_Pause]
+    and     r0, r2
+    cmp     r0, #00h
+    beq     @@return
+    bl      ProcessPauseButtonPress
+    cmp     r0, #FALSE
+    beq     @@return
+    ldr     r0, =SubGameMode1
+    ldrh    r1, [r0]
+    add     r1, #01h
+    strh    r1, [r0]
+    ldr     r0, =NonGameplayFlag
+    mov     r1, #NonGameplayFlag_PauseScreen
+    strb    r1, [r0]
+
+@@return:
+    ldr     r1, =@ReturnFromNoClipCheck
+    bx      r1
+    .pool
+
+    .align 2
+@DebugMenuModifyHealthAndAmmoHighjack:
+@@check_max_energy:
+    cmp     r3, #06h
+    bne     @@check_metroids
+    ldr     r5, =SamusUpgrades+SamusUpgrades_MaxEnergy
+    sub     r2, r5, #02h
+    mov     r8, r2
+    mov     r1, #03h    ; length of numbers you can choose left/right
+    mov     r7, #02h    ; what is r7?
```

Comment:
i think its supposed to be on whether you're changing the current or the maximum

Since modify current health sets it to 1, but max health sets it to 2
Then current missiles sets it to 1 and max missiles sets it to 2
And current PBs set it to 3, but max pbs set it to 4

And then based on some other conditions somehow the return value is calculated.